isic/core/api/collection.py

```diff
@@ -1,105 +1,126 @@
 from django.contrib import messages
-from django.contrib.auth.models import User
 from django.http.response import JsonResponse
-from django.utils.decorators import method_decorator
-from django_filters.rest_framework import DjangoFilterBackend
-from drf_yasg.utils import swagger_auto_schema
-from rest_framework import status
-from rest_framework.decorators import action
-from rest_framework.exceptions import PermissionDenied
-from rest_framework.response import Response
-from rest_framework.viewsets import ReadOnlyModelViewSet
+from django.shortcuts import get_object_or_404
+from ninja import Field, ModelSchema, Router, Schema
+from ninja.pagination import paginate
+from pydantic.types import conlist, constr
 
+from isic.core.constants import ISIC_ID_REGEX
 from isic.core.models.collection import Collection
-from isic.core.permissions import IsicObjectPermissionsFilter
-from isic.core.serializers import CollectionSerializer, IsicIdListSerializer, SearchQuerySerializer
+from isic.core.pagination import CursorPagination
+from isic.core.permissions import get_visible_objects
+from isic.core.serializers import SearchQueryIn
 from isic.core.services.collection.image import (
     collection_add_images_from_isic_ids,
     collection_remove_images_from_isic_ids,
 )
 from isic.core.tasks import populate_collection_from_search_task
 
-from .exceptions import Conflict
+router = Router()
 
 
-@method_decorator(
-    name="list", decorator=swagger_auto_schema(operation_summary="Return a list of collections.")
+class CollectionOut(ModelSchema):
+    class Config:
+        model = Collection
+        model_fields = ["id", "name", "description", "public", "pinned", "locked", "doi"]
+
+    doi_url: str | None = Field(alias="doi_url")
+
+
+@router.get("/", response=list[CollectionOut], summary="Return a list of collections.")
+@paginate(CursorPagination)
+def collection_list(request, pinned: bool | None = None) -> list[CollectionOut]:
+    queryset = get_visible_objects(request.user, "core.view_collection", Collection.objects.all())
+    if pinned is not None:
+        queryset = queryset.filter(pinned=pinned)
+    return queryset
+
+
+@router.get("/{id}/", response=CollectionOut, summary="Retrieve a single collection by ID.")
+def collection_detail(request, id: int) -> CollectionOut:
+    qs = get_visible_objects(request.user, "core.view_collection", Collection.objects.all())
+    collection = get_object_or_404(qs, id=id)
```

Comment:
This looks exactly like their [documented example](https://django-ninja.rest-framework.com/tutorial/other/crud/), and that code does not explicitly declare the 404 response type.

Reply:
Yeah, I suppose it wouldn't make much sense since a 404 could be on every endpoint. I see now where they handle the exception mapping https://django-ninja.rest-framework.com/guides/errors/?h=exceptions#override-the-default-exception-handlers.